### PR TITLE
Branches - Eve: Solar System

### PR DESCRIPTION
**Pull request by tofuandeve**
# Solar System
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Question 	|  Answer
|---	|---	|
| When does the `initialize` method run? What does it do? | The 'initialize' method run whenever we construct an instance of a class by calling .new. It initialize the instance variables with values being passed in (if needed)
| Why do you imagine we made our instance variables readable but not writable? | Making instance variables readable so that we can check the values stored in those variables of an instance of a class. Making it not writable to prevent people from changing values of instance variables. An example could be: a class AdaStudent has instance variable named @ssn. Making @ssn readable so that one can use this SSN to fill out Ada's paperwork, but it's important to make this @ssn not writable to prevent Ada from changing its value to "hello this is Ada"
| How would your program be different if each planet was stored as a `Hash` instead of an instance of a class? | It would be hard to keep track of the planets when there are too many planets also looking up and modifying a specific planet in these planets would be much harder. And readability might also be a problem
| How would your program be different if your `SolarSystem` class used a `Hash` instead of an `Array` to store the list of planets? |It depends on how I structure my hash. But the way I iterate through @planets to gather information of planets will be different.
| There is a software design principle called the SRP. The Single Responsibility Principle (SRP) says that each class should be responsible for exactly one thing. Do your classes follow SRP? What responsibilities do they have? | Planet class is responsible for constructing and managing information of each Planet instance. SolarSystem is responsible for constructing and managing in formation of each SolarSystem instance
| How did you organize your `require` statements? Which files needed `require`s, and which did not? What is the pattern? | solar_system.rb has Planet instances there for it needs to know about planet.rb ('require' file planet.rb). main.rb uses solar_system.rb which knows about Planet, therefore, only file solar_system.rd is required in main.rb. Test files: solar_system_test.rb test SolarSystem class, file solar_system.rb is required. planet_test.rb test Planet class, file planet.rb is required.

### CODE REVIEW

**Comment by tildeee** [comment]
# Solar System
## What We're Looking For

Feature | Feedback
--- |---
**Baseline** | 
Whitespace (indentation, vertical space, etc) | x
Variable names | x
Git hygiene | x
**Planet**  |
`initialize` method stores parameters as instance variables with appropriate reader methods | x
`summary` method returns a string | x
**SolarSystem** |
`initialize` creates an empty list of planets | x
`add_planet` takes an instance of `Planet` and adds it to the list | x
`list_planets` returns a string | x
`find_planet_by_name` returns the correct instance of `Planet` | x
**CLI** |
Can list planets and quit | x
Can show planet details | x
Can add a planet | x
Complex functionality is broken out into separate methods | x
**Overall** |



Eve!!! Your Solar System project is fantastic!!! Your code is thorough and logical. It's clean and covers all of the angles. I'm particularly really thrilled by all of the methods you pulled out into `main.rb` and how all of the code reused a lot of those helper methods. Also, I'm really excited by your tests-- they're the exact level of detail and coverage I'm looking for.


Your code style is also so nice to read; it's dense, but also so concise and elegant. Your code makes me happy to read!


I have a couple of comments and suggestions for minor areas of improvement.


That being said, your project is great overall. Keep up the good work!

File: lib/solar_system.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    planets_list = "Planets orbiting #{@star_name}\n"
+    @planets.length.times do |index|
+      planets_list << "#{index + 1}. #{planets[index].name.capitalize}\n"
+    end
+    return planets_list
+  end
+  
+  def find_planet_by_name(planet_name)
+    raise ArgumentError.new("Planet name must be a string") if !(planet_name.instance_of? String)
+    planets_by_name = @planets.select {|obj| obj.name == planet_name.capitalize}
```
Comment (tildeee) [inline lib/solar_system.rb:27]: Nice!

Here's a possible refactoring: using `find` instead of `select`. It will give back the element
 that matches first. If there is no match, then it will give back `nil`

File: lib/solar_system.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    planet = (planets_by_name.empty?) ? nil : planets_by_name[0]
+    return planet
+  end
+  
+  def distance_between(planet_name1, planet_name2)
+    raise ArgumentError.new("Arguments must be strings") if !(planet_name1.instance_of? String) || !(planet_name2.instance_of? String)
+    first_planet = find_planet_by_name(planet_name1)
+    second_planet = find_planet_by_name(planet_name2)
```
Comment (tildeee) [inline lib/solar_system.rb:35]: Great work re-using the other instance methods!

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+def get_positive_number_from_user
+  input = Integer(gets.chomp) rescue false
+  while !input || input <= 0
+    print "Invalid input. Input must be a positive number: "
+    input = Integer(gets.chomp) rescue false
+  end 
+  return input  
+end
+
+def get_string_input_from_user()
```
Comment (tildeee) [inline main.rb:12]: Really thorough and thoughtful logic in here. It anticipates a lot of interesting things users can do. Well done

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  print "Please enter in your planet's distance from the sun: "
+  distance = get_positive_number_from_user
+  
+  print "Please enter in your planet's fun fact: "
+  fact = get_string_input_from_user
+  
+  return Planet.new(name: name, color: color, mass_kg: mass_kg, distance: distance, fact: fact)
+end
+
+def get_2_planet_names_from_user(solar_system)
```
Comment (tildeee) [inline main.rb:60]: I love how your code looks in this method! It looks so nice!!

File: tests/planet_test.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
+require 'minitest/autorun'
+require 'minitest/reporters'
+require 'minitest/skip_dsl'
+
+require_relative '../lib/planet'
+
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe "Planet class" do
```
Comment (tildeee) [inline tests/planet_test.rb:9]: Really really wonderful tests! They're really thorough and cover all of the important parts. Keep it up!

File: tests/planet_test.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  describe "Constructor" do
+    it "Error checks input being passed in for mass_kg" do
```
Comment (tildeee) [inline tests/planet_test.rb:11]: Minor nitpick: It may be helpful to include in the test name that this error happens for negative mass_kg (same as the next distance test)
